cli/bptest/list.go

```diff
@@ -0,0 +1,129 @@
+package bptest
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/discovery"
+	testing "github.com/mitchellh/go-testing-interface"
+)
+
+const (
+	discoverTestFilename = "discover_test.go"
+)
+
+type bpTest struct {
+	name     string
+	config   string
+	location string
+}
+
+// getTests returns slice of all blueprint tests
+func getTests(intTestDir string) ([]bpTest, error) {
+	tests := []bpTest{}
+	discoveredTests, err := getDiscoveredTests(intTestDir)
+	if err != nil {
+		return nil, err
+	}
+	tests = append(tests, discoveredTests...)
+
+	explicitTests, err := getExplicitTests(intTestDir)
+	if err != nil {
+		return nil, err
+	}
+	tests = append(tests, explicitTests...)
+
+	return tests, nil
+}
+
+// getDiscoveredTests returns slice of discovered blueprint tests
+func getDiscoveredTests(intTestDir string) ([]bpTest, error) {
+	discoverTestFile, err := getDiscoverTestFile(intTestDir)
+	// skip discovering tests if no discoverTestFile
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, err
+		}
+		Log.Warn(fmt.Sprintf("Skipping discovered test. %s not found.", discoverTestFilename))
+	}
+	// if discoverTestFile is present, find auto discovered tests
+	tests := []bpTest{}
+	if discoverTestFile != "" {
+		discoverTestName, err := getDiscoverTestName(discoverTestFile)
+		if err != nil {
+			return nil, err
+		}
+		discoveredSubTests := discovery.FindTestConfigs(&testing.RuntimeT{}, intTestDir)
+		for testName, fileName := range discoveredSubTests {
+			tests = append(tests, bpTest{name: fmt.Sprintf("%s/%s", discoverTestName, testName), config: fileName, location: discoverTestFile})
+		}
+	}
+	sort.SliceStable(tests, func(i, j int) bool { return tests[i].name < tests[j].name })
+	return tests, nil
+}
+
+func getExplicitTests(path string) ([]bpTest, error) {
+	// find dirs within test/integration
+	dirs, err := findDirContentsFilter(path, func(f fs.FileInfo) bool { return f.IsDir() })
```

Comment:
Not sure we should add a new function here? Any reason not to just use a glob? https://pkg.go.dev/path/filepath#Glob

`**/*_test.go` should pick up all test files.

Reply:
It looks like globstar is not supported https://github.com/golang/go/issues/11862 for `filepath.Glob`
Previously this seemed to worked as tests were always found from `test/integration/explicitName/*_test.go` and `**/*_test.go` was interpreted used as `*/*_test.go`.